text/3305-static-in-pattern.md

```diff
@@ -0,0 +1,285 @@
+
+```rust
+// libc crate
+
+cfg_if! {
+    if #[cfg(target_env = "cosmopolitan")] {
+        extern "C" {
+            pub static EINVAL: i32;
+            pub static ENOSYS: i32;
+            pub static ENOENT: i32;
+        }
+    } else {
+        pub const EINVAL: i32 = 42;
+        pub const ENOSYS: i32 = 43;
+        pub const ENOENT: i32 = 44;
+    }
+}
+
+// stdlib code
+
+use libc::*;
+
+fn process_error(error_code: i32) {
+    match error_code {
+        // Compiler throws error EO530 on Cosmopolitan,
+        // because `static`s can't be used in patterns, only `const`s
+        EINVAL => do_stuff(),
+        ENOSYS => panic!("oh noes"),
+        ENOENT => make_it_work(),
+        _ => do_different_stuff(),
+    }
+}
+```
+
+Because Rust patterns don't support statics, all the `match` expressions in the standard library that refer to POSIX constants would currently need to be rewritten to accommodate Cosmopolitan.
+
+```rust
+// stdlib code adapted for cosmopolitan
+
+use libc::*;
+
+fn process_error(error_code: i32) {
+    if error_code == EINVAL {
+        do_stuff();
+    } else if error_code == ENOSYS {
+        panic!("oh noes");
+    } else if error_code == ENOENT {
+        make_it_work();
+    } else {
+        do_different_stuff();
+    }
+}
+```
+
+Needless to say, this is unlikely to ever be upstreamed. Allowing statics in patterns would solve this use-case much more cleanly.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Rust patterns can refer to constants.
+
+```rust
+const EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+With this feature, they can refer to statics as well.
+
+```rust
+static EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+Mutable statics are not allowed, however. Patterns can't reference information that can change at runtime, and also can't be `unsafe`.
+
+```rust
+
+static mut EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        // ERROR can't refer to mutable statics in patterns
+        /* EVERYTHING => println!("have all of it"), */
+        _ => println!("need moar"),
+    }
+}
+```
+
+Statics from `extern` blocks are allowed, but they must be marked as trusted using the (not-yet-implemented) [trusted external statics](https://github.com/rust-lang/lang-team/issues/149) feature.
+
+```rust
+extern "C" {
+    #[unsafe(trusted_extern)]
+    static EVERYTHING: i32;
+}
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+For a static to be eligible for use in a pattern, it must:
```

Comment:
I am not sure if this is enough for soundness. The reason `const` are not allowed to even point to immutable `static` is pattern matching soundness: if you have a `const` of type `&i32`, we better make sure that tie integer itself is truly immutable. But if you define a `static X: &i32` by doing `&mut MUTABLE_STATIC` and then transmuting that, and then you do `const C: &i32 = X`, now you have yourself a potentially unsound pattern.

But for `extern static` we don't know their values, so we can't know if they do shenanigans like that. Also for all we know their type must be a lie. There is an entire lion's den of issues here that we have so far worked very hard to avoid (e.g. grep the compiler source code for `can_access_statics` and read all the comments associated with it), and all of this work is for naught if we allow `extern static` in patterns.

Reply:
Yeah, I hope we can fix this in a future edition

Reply:
Not sure this needs a new edition, we just need to be okay with semver hazards and post-monomorphization errors.

But anyway, that won't help for `extern static`. I can see some ways of resolving the structural match situation that would probably also make other kinds of "opaque consts" like `extern static` be fine (*), but as of now I don't know where the structural match stuff will go so I'd rather not rely on it going any particular way.

(*) Well, there are also the issues around "evading match arms" by changing the pattern mid-evaluation, but that can probably be avoided by not doing any exhaustiveness checking and clearly documenting when opaque patterns vs valtree patterns are used (because they will probably *not* always be equivalent).

Reply:
Sorry, I meant specifically that with `extern static` (non-mut) I would like declaring them to be the unsafe part, and it would require precise time and truly immutable memory and such, then the accesses themselves can be the safe part and people can worry a lot less.

Doesn't help the actual valtree matching (or not) part of it though.

Reply:
The condition we need here is that *any pointer we reach through this memory points to immutable memory, recursively*. This is required for matching on references. It is why I keep saying this is a bigger soundness issue than the RFC lets on.

I doubt trusted statics want to require *that*.

Reply:
Ah, because the match always follows recursively, but use of a static might not follow recursively?

Reply:
Yes.

Reply:
Aren't there many other ways to trigger UB with a static that lies about it's type? The match code in my PR just does `PartialEq::eq` to check if the scrutinee matches for all non-primitive types, so the safety requirement of the match should be no greater than that of that safe-to-call method.

Reply:
Mutating the value referenced by a `&&&i32` is itself UB imho, because you're mutating a shared reference without `UnsafeCell`...so rustc can safely assume the second `STATIC` arm can't be reached if it has type `&&&i32`.

Reply:
> but if that value can be valtree'd, then those two things are not UB-equivalent.

How? The current stuctural match rules oly allow constants (and, with this feature, statics) if `PartialEq::eq` is guaranteed to give the same results as a valtree-based match.

Reply:
> Mutating the value referenced by a &&&i32 is itself UB imho,

That's probably not correct. You certainly shouldn't rely on it. The rules for UB on references will probably not recursively follow references to arbitrary depth, but instead work on a 'shallow' level -- mutation is only UB one level down.

As a concrete example, [Miri sees no UB here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7a6973f3fdc94a0bff266eabebb299b8).

> How? The current stuctural match rules oly allow constants (and, with this feature, statics) if PartialEq::eq is guaranteed to give the same results as a valtree-based match.

The current rules are a mess and before anyone relies on them for anything, we need to re-do those rules. See https://github.com/rust-lang/rust/issues/74446.

The future rules will, hopefully, give users the ability to `unsafe impl StructuralEq for MyType`. (And in fact they can already do that, even without `unsafe`, by using unstable features.)

Users of course cannot expect things to behave completely sanely when they do that, but we still need to specify precisely what happens. Note in particular that saying "it is UB to have such an instance" does not work, that's simply not how we are using UB in Rust -- UB can only arise as a consequence of the program executing a statement in the wrong way, and only under conditions that can actually be decided automatically.  "This type has a PartialEq that behavesd equivalently to valtrees" is not a decidable question (halting problem, yada yada), and hence cannot be grounds for UB.

Reply:
> "This type has a PartialEq that behaves equivalently to valtrees" is not a decidable question

But "at the point this specific match statement was executed, this specific call to `PartialEq::eq` completed and returned a value without panicking, and that return value was different from what a valtree-based match would have returned" is decidable, no?

(This would mean `loop {}` is a valid `PartialEq::eq` impl, but I don't think that's a problem).

Reply:
>> Mutating the value referenced by a &&&i32 is itself UB imho,
> That's probably not correct. You certainly shouldn't rely on it.

All the issues that arise with "a static of type `&&&i32` could change during a pattern match" already occur with "a scrutinee of type `&&&i32` could change during a pattern match", no?

Reply:
> But "at the point this specific match statement was executed, this specific call to PartialEq::eq completed and returned a value without panicking, and that return value was different from what a valtree-based match would have returned" is decidable, no?

So now we need a way to express that in MIR, and carefully specify exactly what that does and when.

That might be possible, but it'll be a lot of work, and it'll also make the core language spec more complicated for a niche feature.

Instead I think we should just specify that statics are treated as opaque for pattern matching and always use `PartialEq` no matter their type. We don't have any other "UB due to valtree inconsistencies", and IMO there is no good reason to introduce such UB here.

> All the issues that arise with "a static of type &&&i32 could change during a pattern match" already occur with "a scrutinee of type &&&i32 could change during a pattern match", no?

No, patterns and scrutinees are treated very differently in pattern matching, because the pattern is inherently known at compile-time and the scrutinee isn't. This is not a symmetric situation. This RFC is proposing to make it more symmetric, but that is exactly why this is a bug fundamental change to pattern matching.

See [here](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-lang.2Fwg-unsafe-code-guidelines/topic/SB.20UnsafeCell.20.2B.20exhaustiveness.20check.20.2B.20layout.20opts) for a long discussion about how to ensure that we have UB when the scrutinee changes during pattern matching. The same approach does not obviously work for patterns, since it relies heavily on the fact that we compare the same value X against a number of things -- Rust has some ingredients that make it possible to ensure that X doesn't change when it is used multiple times. But with patterns that just happen to use the same extern static in several places across several patterns, we have none of that structure, so it'd be even more complicated.

So, again, we should just specify that extern statics are treated as opaque, and we should make no claims that this is equivalent to a const of the same type and value.